File: Java/Recover-Rotated-Sorted-Array.java
public class Solution {
    /**
     * @param nums: The rotated sorted array
     * @return: void
     */
    public void swap(ArrayList<Integer> nums, int left, int right) {
        int temp;
        while (left < right) {
            temp = nums.get(left);
            nums.set(left, nums.get(right));
            nums.set(right, temp);
            left ++;
            right --;
        }
    }
    public void recoverRotatedSortedArray(ArrayList<Integer> nums) {
        // write your code
        int cur = -1;
        for (int i = 1; i < nums.size(); i ++) {
            if (nums.get(i) < nums.get(i - 1)) {
                cur = i;
                break;
            }
        }
        if (cur == -1) return;
        swap(nums, 0, cur - 1);
        swap(nums, cur, nums.size() - 1);
        swap(nums, 0, nums.size() - 1);
    }
}

File: Python/Recover-Rotated-Sorted-Array.py
class Solution:
    """
    @param nums: The rotated sorted array
    @return: nothing
    """
    def reverse(self, nums, left, right):
        while left < right:
            temp = nums[left]
            nums[left] = nums[right]
            nums[right] = temp
            left += 1
            right -= 1

    def recoverRotatedSortedArray(self, nums):
        # write your code here
        pos = -1
        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                pos = i
                break
        if (pos == -1): return
        ##reversed(nums, 0, pos)
        self.reverse(nums, 0, pos)
        self.reverse(nums, pos + 1, len(nums) - 1)
        self.reverse(nums, 0, len(nums) - 1)